File: TopCoder/python/misc/BaseballGame.py
class Solution:
    def calPoints(self, ops):
        list = []
        result = 0
        for op in ops:
            if op == '+':
                value = 0
                if len(list) > 0:
                    value = value + list[len(list) - 1]

                if len(list) > 1:
                    value = value + list[len(list) - 2]

                result = result + value
                list.append(value)
            elif op == 'D':
                value = 0
                if len(list) > 0:
                    value = list[len(list)-1]*2
                result = result + value
                list.append(value)
            elif op == 'C':
                if len(list) > 0:
                    value = list.pop()
                    result = result - value

            else:
                result = result + int(op)
                list.append(int(op))
        return result

ops = ["5","-2","4","C","D","9","+","+"]
sol = Solution()
res = sol.calPoints(ops)
print(res)

File: TopCoder/src/main/java/com/leetcode/linkedlist/BaseballGame.java
package com.leetcode.linkedlist;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.IntStream;

//https://leetcode.com/problems/baseball-game/
public class BaseballGame {
    public int calPoints(String[] ops) {
        int result = 0;
        List<Integer> list = new LinkedList<>();
        for (String str : ops) {
            try {
                int rt = Integer.parseInt(str);
                result += rt;
                list.add(rt);
            } catch (Exception e) {
                if ("+".equals(str)) {
                    int value = 0;
                    if (list.size() > 0) {
                        value += list.get(list.size()-1);
                    }
                    if (list.size() > 1) {
                        value += list.get(list.size()-2);
                    }
                    result += value;
                    list.add(value);
                } else if ("D".equals(str)) {
                    if (list.size() > 0) {
                        int value = list.get(list.size()-1);
                        result += value*2;
                        list.add(value*2);
                    }
                } else if ("C".equals(str)) {
                    if (list.size() > 0) {
                        int value = list.get(list.size()-1);
                        result -= value;
                        list.remove(list.size()-1);
                    }
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        String[] str = {"5","-2","4","C","D","9","+","+"};
        int result = new BaseballGame().calPoints(str);
        System.out.println(result);
    }
}
